libcudacxx/include/cuda/__cmath/fast_modulo_division.h

```diff
@@ -0,0 +1,181 @@
+ * Extract higher bits after multiplication
+ **********************************************************************************************************************/
+
+template <typename _Tp, typename _Up>
+[[nodiscard]] _CCCL_API constexpr _CUDA_VSTD::common_type_t<_Tp, _Up> __multiply_extract_higher_bits(_Tp __x, _Up __y)
+{
+  using namespace _CUDA_VSTD;
+  static_assert(__cccl_is_integer_v<_Tp> && sizeof(_Tp) <= 8, "unsupported type");
+  static_assert(__cccl_is_integer_v<_Up> && sizeof(_Up) <= 8, "unsupported type");
+  if constexpr (is_signed_v<_Tp>)
+  {
+    _CCCL_ASSERT(__x >= 0, "__x must be non-negative");
+    _CCCL_ASSUME(__x >= 0);
+  }
+  if constexpr (is_signed_v<_Up>)
+  {
+    _CCCL_ASSERT(__y >= 0, "__y must be non-negative");
+    _CCCL_ASSUME(__y >= 0);
+  }
+  constexpr auto __mul_bits = ::cuda::next_power_of_two(__num_bits_v<_Tp> + __num_bits_v<_Up>);
+  using __larger_t          = __make_nbit_uint_t<__mul_bits>;
+  using __ret_t             = common_type_t<_Tp, _Up>;
+  auto __ret                = (static_cast<__larger_t>(__x) * __y) >> (__mul_bits / 2);
+  return static_cast<__ret_t>(__ret);
+}
+
+/***********************************************************************************************************************
+ * Fast Modulo/Division based on Precomputation
+ **********************************************************************************************************************/
+
+template <typename _Tp, bool _IsDivisorNotOne = false>
+class fast_mod_div
+{
+  static_assert(_CUDA_VSTD::__cccl_is_integer_v<_Tp> && sizeof(_Tp) <= 8, "unsupported type");
+  using __unsigned_t = _CUDA_VSTD::make_unsigned_t<_Tp>;
+
+public:
+  fast_mod_div() = delete;
+
+  _CCCL_API explicit fast_mod_div(_Tp __divisor1) noexcept
+      : __divisor{__divisor1}
+  {
+    using namespace _CUDA_VSTD;
+    using __larger_t = __make_nbit_uint_t<__num_bits_v<_Tp> * 2>;
+    _CCCL_ASSERT(!_IsDivisorNotOne || __divisor1 != 1, "divisor must not be one");
+    if constexpr (is_signed_v<_Tp>)
+    {
+      __shift      = ::cuda::ceil_ilog2(__divisor1) - 1; // is_pow2(x) ? log2(x) - 1 : log2(x)
+      auto __k     = __num_bits_v<_Tp> + __shift;
+      __multiplier = ::cuda::ceil_div(__larger_t{1} << __k, __divisor);
+    }
+    else
+    {
+      __shift = ::cuda::ilog2(__divisor1);
+      if (::cuda::is_power_of_two(__divisor))
+      {
+        __multiplier = 0;
+        return;
+      }
+      auto __k              = __num_bits_v<_Tp> + __shift;
+      __multiplier          = ((__larger_t{1} << __k) + (__larger_t{1} << __shift)) / __divisor;
+      auto __multiplier_low = (__larger_t{1} << __k) / __divisor;
+      __add                 = (__multiplier_low == __multiplier);
+    }
+  }
+
+  template <typename _Up>
+  [[nodiscard]] _CCCL_API friend _CUDA_VSTD::common_type_t<_Tp, _Up>
+  operator/(_Up __dividend, fast_mod_div<_Tp> __divisor1) noexcept
+  {
+    using namespace _CUDA_VSTD;
```

Comment:
We cannot do that, we need to always fully qualify

Reply:
those are fine

Reply:
However, pulling in all namespaces cuda, cuda::std, cuda::std::ABI, seems a bit wasteful when we could just qualify things